### PR TITLE
Some additional work around enum BETWEEN constraints 

### PR DESCRIPTION
**Pull request by roji**
@celluj34 this continues on your work in #96, would you like to give this a look? The PR is split into 3 commits for easier reviewing, I'd be happy for any feedback.

### CODE REVIEW

File: EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs
Comment (celluj34) [inline EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs:129]: I've always been more of a fan of `.ToList()`, but that's just preference.

Comment (roji) [reply EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs:129]: I tend to prefer ToArray when I know the length is fixed and I won't need to add/remove elements... But yeah, it doesn't really matter.

Any other comments on this PR?

Comment (celluj34) [reply EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs:129]: Well this is weird, maybe I'm not used to Github's interface, but I had a few other comments but I can only click "Add Review Comment" instead of a normal comment. Perhaps that means something different than what I thought. I'll make actual them actual comments so you can see them.

**Comment by celluj34** [comment]
I don't know if you could see my other comment, but it's a reply to the `ToList` convo: 

Well this is weird, maybe I'm not used to Github's interface, but I had a few other comments but I can only click "Add Review Comment" instead of a normal comment. Perhaps that means something different than what I thought. I'll make actual them actual comments so you can see them. They all have a "Pending" tag on them:

image

File: EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs
Comment (celluj34) [inline EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs:126]: Can this be `T[]`?

File: EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs
Comment (celluj34) [reply EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs:129]: Oh I think I figured it out!

File: EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs
Comment (roji) [reply EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs:111]: Note that we only look up the method if a new enum underlying type is encountered, i.e. only when we execute MakeGenericMethod as well. And the actual lookup of the method takes generally less than MakeGenericMethod.

So we could also cache the open generic method (before MakeGenericMethod), there's just not much point in doing that.

File: EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs
Comment (roji) [reply EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs:126]: I don't think so - the actual type of the array is `object[]` - it's created above with `Cast<object>().Distinct().ToArray()`. That cannot be just cast to e.g. `int[]`, which is why we must cast the individual members via `Cast<T>()` inside.

File: EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs
Comment (roji) [reply EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs:136]: Good idea! Will do that.

**Comment by roji** [comment]
Did your proposed change @celluj34, will wait on further feedback before merging (if you're interested, I can also ping you on other PRs here).

**Comment by celluj34** [comment]
Everything looks good to me! Sure, if you've got a few more I can take a look.

File: EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs
Comment (celluj34) [reply EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs:126]: It _would_ feel better if we were able to, but then we'd have to start working around the non-generic Array type and that's not very much fun.

**Review by celluj34 — APPROVED** [review]

File: EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs
Comment (roji) [reply EFCore.CheckConstraints/Internal/EnumCheckConstraintConvention.cs:126]: Yep.

**Pull request merged**
